### PR TITLE
feat: do not instrument npm or yarn

### PR DESCRIPTION
**Pull request by basti1302**
If a process is started with `npm start` or `yarn start`, two separate
Node.js processes will be started. The first is `node /path/to/npm`
(or `node /path/to/yarn`). This process will kick off another process
with the actual application (for example, `node /usr/src/app`).

With this change, even when NODE_OPTIONS is set to
"--require /usr/src/app/node_modules/@instana/collector/src/immediate",
only the actual application will be instrumented, but not the npm/yarn
process.

### CODE REVIEW

File: packages/collector/dummy-app/package-lock.json [hunk 1]
```diff
@@ -5,36 +5,36 @@
   "requires": true,
   "dependencies": {
     "@instana/autoprofile": {
-      "version": "1.100.1",
-      "resolved": "https://registry.npmjs.org/@instana/autoprofile/-/autoprofile-1.100.1.tgz",
-      "integrity": "sha512-IvCQsyUyN0nkfHbVwEYTtXw42+9DEcK72lLmHN7womrp33ehRnRevCWE/MuJxNaPgnTVbHpeDzVexGfbPw+MiQ==",
+      "version": "1.110.5",
```
Comment (basti1302) [inline packages/collector/dummy-app/package-lock.json:8]: The changes in this file are unrelated to the actual change. I just updated the Instana package to latest for the test application.

File: packages/collector/test/immediate/test.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+'use strict';
```
Comment (basti1302) [inline packages/collector/test/immediate/test.js:1]: This is the integration test for the change in `@instana/collector`. It starts an application under test via `npm start` and verifies that only the actual application under test is instrumented.

File: packages/collector/test/immediate/package.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "immediate",
+  "version": "1.0.0",
+  "description": "test application",
+  "license": "MIT",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/instana/nodejs-sensor.git"
+  },
+  "main": "../apps/express",
```
Comment (basti1302) [inline packages/collector/test/immediate/package.json:10]: TODO: This doesn't make sense yet. We need an uninstrumented app here, and ../apps/express is instrumented manually.

File: packages/core/src/util/excludedFromInstrumentation.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+'use strict';
```
Comment (basti1302) [inline packages/core/src/util/excludedFromInstrumentation.js:1]: This file contains the main logic for excluding process from auto instrumentation.

File: packages/collector/test/immediate/test.js [hunk 5]
```diff
@@ -30,4 +34,59 @@ describe('collector/src/immediate', function() {
       expect(pidsThatContactedTheAgent[controls.getPid()]).to.exist;
     });
   });
+
+  describe('exclude processes from auto instrumentation', () => {
```
Comment (basti1302) [inline packages/collector/test/immediate/test.js:38]: This is the integration test for the change in `@instana/collector`. It starts an application under test via `npm start` and verifies that only the actual application under test is instrumented.

File: packages/core/src/util/excludedFromInstrumentation.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+// If a process is started with `npm start` or `yarn start`, two separate processes will be started. The first is
+// `node /path/to/npm` (or `node /path/to/yarn`). This process will kick off another process with the actual application
+// (for example, `node /usr/src/app`). When
+// NODE_OPTIONS="--require /usr/src/app/node_modules/@instana/collector/src/immediate" is set, we would instrument both
+// processes, that is, npm/yarn as well as the actual application. Attempting to instrument the npm or yarn process has
+// no value and also creates confusing log output, so we exclude them here explicitly.
+const excludeList = [
```
Comment (willianpc) [inline packages/core/src/util/excludedFromInstrumentation.js:9]: As a personal preference, I would use "or", so the array would no longer be necessary. Unless I missed something:
```javascript
const excludePattern = /^.*\/(yarn|npm)$/;
```

Also, I added `$` to restrict the pattern to the exact name `npm` and `yarn`. The way it is, something like `yarna` or `npmagic` will not pass.

Hope it makes sense!

File: packages/google-cloud-run/src/index.js [hunk 7]
```diff
@@ -1,10 +1,15 @@
 'use strict';
 
+const { util: coreUtil } = require('@instana/core');
 const { environment: environmentUtil } = require('@instana/serverless');
 
-environmentUtil.validate();
+let isExcludedFromInstrumentation = coreUtil.excludedFromInstrumentation && coreUtil.excludedFromInstrumentation();
```
Comment (willianpc) [inline packages/google-cloud-run/src/index.js:6]: Same as before, `const` should be enough :)

**Review by willianpc — APPROVED** [review]
LGTM.
Most comments are minimum stuff, except the regex that may need some closer look.
Nevertheless I am approving the PR.

Cheers!

File: packages/core/src/util/excludedFromInstrumentation.js (hunk 6 above)
Comment (basti1302) [reply packages/core/src/util/excludedFromInstrumentation.js]: Yep, that absolutely makes sense. I've changed that accordingly.

File: packages/google-cloud-run/src/index.js (hunk 7 above)
Comment (basti1302) [reply packages/google-cloud-run/src/index.js]: Absolutely. Instead of changing that one place, I added a new eslint rule for similar cases and ran eslint --fix to quickyl fix all unnecessary usages of let throughout the whole code base. There were a bunch.
